servo-tomcat/src/main/java/com/netflix/servo/publish/tomcat/TomcatPoller.java

```diff
@@ -0,0 +1,159 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.netflix.servo.publish.tomcat;
+
+import com.netflix.servo.Metric;
+import com.netflix.servo.annotations.DataSourceType;
+import com.netflix.servo.publish.BaseMetricPoller;
+import com.netflix.servo.tag.BasicTagList;
+import com.netflix.servo.tag.Tag;
+import com.netflix.servo.tag.TagList;
+import com.netflix.servo.tag.Tags;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.management.Attribute;
+import javax.management.AttributeList;
+import javax.management.JMException;
+import javax.management.MBeanServer;
+import javax.management.ObjectName;
+import java.lang.management.ManagementFactory;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Set;
+
+/**
+ * Fetch Tomcat metrics from JMX.
+ */
+public class TomcatPoller extends BaseMetricPoller {
+    private static final List<Metric> EMPTY_LIST = Collections.emptyList();
+    private static final Logger LOGGER = LoggerFactory.getLogger(TomcatPoller.class);
+
+    private static Tag catalina = Tags.newTag("source", "Catalina");
+
+    private static Metric toMetric(long t, ObjectName name, Attribute attribute, Tag dsType) {
+        Tag pool = Tags.newTag("pool", name.getKeyProperty("name"));
+        Tag clazz = Tags.newTag("class", name.getKeyProperty("type"));
+        TagList list = BasicTagList.of(catalina, pool, clazz, dsType);
+        return new Metric(attribute.getName(), list, t, attribute.getValue());
+    }
+
+    private static Metric toGauge(long t, ObjectName name, Attribute attribute) {
+        return toMetric(t, name, attribute, DataSourceType.GAUGE);
+    }
+
+    private static Metric toCounter(long t, ObjectName name, Attribute attribute) {
+        return toMetric(t, name, attribute, DataSourceType.COUNTER);
+    }
+
+    private static final String[] THREAD_POOL_ATTRS = new String[]{
+            "maxThreads",
+            "currentThreadCount",
+            "currentThreadsBusy",
+            "backlog"
+    };
+
+    private static final String[] GLOBAL_REQ_ATTRS = new String[]{
+            "requestCount",
+            "errorCount",
+            "bytesSent",
+            "processingTime",
+            "maxTime"
+    };
+
+    private static final String[] EXECUTOR_ATTRS = new String[]{
+            "poolSize",
+            "activeCount"
+    };
+
+    private static void fetchRequestProcessorMetrics(long now, MBeanServer mbs,
+                                                     List<Metric> metrics) throws JMException {
+        final ObjectName globalName = new ObjectName("Catalina:type=GlobalRequestProcessor,*");
```

Comment:
It is probably worth noting that in order for this to match the service name needs to be set in server.xml:

```xml
<Service name="Catalina">
```

It would be nice if we could auto-detect the set of tomcat services that are available in jmx as users picking a different name has bitten us many times. Unfortunately, because of the way it is modeled I think this would require a full scan of jmx and then looking for something with the right sub-properties on the `ObjectName`.